dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,635 @@
+          }
+        }
+      }
+    }
+    return false;
   }
-}
 
-template <>
-void RandomBBoxCrop<CPUBackend>::RunImpl(SampleWorkspace &ws) {
-  const auto &boxes_tensor = ws.Input<CPUBackend>(0);
+  void Run(SampleWorkspace &ws) {
+    const auto &boxes_tensor = ws.Input<CPUBackend>(0);
+    auto nboxes = boxes_tensor.dim(0);
+    auto ncoords = ndim * 2;
+
+    std::vector<Box<ndim, float>> bounding_boxes(nboxes);
+    ReadBoxes(make_span(bounding_boxes),
+              make_cspan(boxes_tensor.data<float>(), boxes_tensor.size()), bbox_layout_);
+
+    std::vector<int> labels;
+    if (has_labels_) {
+      const auto &labels_tensor = ws.Input<CPUBackend>(1);
+      auto nlabels = labels_tensor.dim(0);
+      DALI_ENFORCE(nlabels == nboxes,
+        make_string("Unexpected number of labels. Expected: ", nboxes, ", got ", nlabels));
+      labels.resize(nlabels);
+      const auto *label_data = labels_tensor.data<int>();
+      for (int i = 0; i < nlabels; i++) {
+        labels[i] = label_data[i];
+      }
+    }
+
+    int sample = ws.data_idx();
+    ProspectiveCrop prospective_crop;
+    while (!prospective_crop.success) {
+      prospective_crop = FindProspectiveCrop(bounding_boxes, labels, sample);
+    }
+
+    WriteCropToOutput(ws, prospective_crop.crop);
+
+    const auto &selected_boxes  = prospective_crop.boxes;
+    WriteBoxesToOutput(ws, selected_boxes);
+
+    if (has_labels_) {
+      const auto &selected_labels = prospective_crop.labels;
+      DALI_ENFORCE(selected_boxes.size() == selected_labels.size(),
+        make_string("Expected boxes.size() == labels.size(). Received: ", selected_boxes.size(),
+          " != ", selected_labels.size()));
+      WriteLabelsToOutput(ws, selected_labels);
+    }
+  }
+
+ private:
+  struct ProspectiveCrop {
+    bool success = false;
+    Box<ndim, float> crop{};
+    std::vector<Box<ndim, float>> boxes;
+    std::vector<int> labels;
+
+    ProspectiveCrop(bool success,
+                    const Box<ndim, float> &crop,
+                    const std::vector<Box<ndim, float>> &boxes,
+                    const std::vector<int> &labels)
+        : success(success), crop(crop), boxes(boxes), labels(labels) {}
+    ProspectiveCrop() = default;
+  };
+
+  const ProspectiveCrop FindProspectiveCrop(const std::vector<Box<ndim, float>> &bounding_boxes,
+                                            const std::vector<int> &labels,
+                                            int sample) {
+    auto &rng = rngs_[sample];
+    std::uniform_int_distribution<> idx_dist(0, sample_options_.size() - 1);
+    SampleOption option = sample_options_[idx_dist(rng)];
+    bool absolute_crop_dims = has_crop_shape_;
+
+    if (option.no_crop) {
+      Box<ndim, float> no_crop = Uniform<ndim>(0.0f, 1.0f);
+      if (absolute_crop_dims) {
+        auto &input_shape = input_shape_[sample];
+        for (int d = 0; d < ndim; d++)
+          no_crop.hi[d] *= input_shape[d];
```

Comment:
Isn't it

```suggestion
        for (int d = 0; d < ndim; d++)
          no_crop.hi[d] = input_shape[d];
```
?

Reply:
It is.

Reply:
I would add a test case for this as this apparently slipped through.